apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.RetryConformanceTests/RetryConformanceTest.cs

```diff
@@ -0,0 +1,370 @@
+    }
+
+    private async Task<TestResponse> CreateRetryTestResource(Method method, JArray instruction)
+    {
+        var stringContent = GetBodyContent(method.Name, instruction);
+        HttpResponseMessage response = await _httpClient.PostAsync("retry_test", stringContent);
+        response.EnsureSuccessStatusCode();
+        var responseMessage = await response.Content.ReadAsStringAsync();
+        return JsonConvert.DeserializeObject<TestResponse>(responseMessage);
+    }
+
+    private void RunRetryTest(TestResponse response, StorageResources resources, bool preConditionsPresent)
+    {
+        AddRetryIdHeader(response.Id);
+
+        // Use method mapping to call the StorageTestBench API.
+        if (_methodMappings.TryGetValue(GetMethodName(response), out var invocation))
+        {
+            if (invocation.ProjectIdRequired)
+            {
+                invocation.ProjectId = ProjectId;
+            }
+
+            if (invocation.BucketNameRequired)
+            {
+                invocation.BucketName = resources.BucketName;
+            }
+
+            if (invocation.ObjectNameRequired)
+            {
+                invocation.ObjectName = resources.ObjectName;
+            }
+
+            if (invocation.HmacKeyRequired)
+            {
+                invocation.BucketName = resources.BucketName;
+                invocation.HmacKey = resources.HmacKey;
+            }
+
+            if (invocation.AccessIdRequired)
+            {
+                invocation.AccessId = AccessId;
+            }
+            if (invocation.MetagenerationRequired)
+            {
+                invocation.Metageneration = 1;
+            }
+            if (invocation.ServiceAccountEmailRequired)
+            {
+                invocation.ServiceAccountEmail = _storageClient.GetStorageServiceAccountEmail(ProjectId);
+            }
+
+            if (invocation.NotificationRequired)
+            {
+                invocation.Notification = resources.Notification;
+            }
+
+            invocation.Invoke(_storageClient, preConditionsPresent);
+        }
+    }
+
+    private StorageClient InitializeClient()
+    {
+        var clientBuilder = new StorageClientBuilder
+        {
+            BaseUri = TestBenchUrl + "storage/v1/"
+        };
+
+        return clientBuilder.Build();
+    }
+
+    private void CreateMethodMapping(RetryTest test)
+    {
+        foreach (var method in test.Methods)
+        {
+            _methodMappings.TryAdd(method.Name, GetMappedFunction(method.Name));
+        }
+    }
+
+    private MethodInvocation GetMappedFunction(string name) => name.ToLowerInvariant() switch
+    {
+        // TODO: Add all mappings from the json file and throw exception in the default case. To be optimised
+
+        "storage.buckets.delete" => new MethodInvocation(s_clientType.GetMethod(nameof(StorageClient.DeleteBucket), new System.Type[] { typeof(string), typeof(DeleteBucketOptions) })) { BucketNameRequired = true },
```

Comment:
Seems clever how all of this has been parameterized!

Reply:
Thanks it was Rishabh's idea to implement it via reflection :)